common/middleware/ensure-authenticated.test.js

```diff
@@ -54,10 +57,129 @@ describe('Authentication middleware', function() {
       })
     })
 
-    context('when the access token has expired', function() {
+    describe('when the access token has expired', function() {
       beforeEach(function() {
         req.session.authExpiry = Math.floor(new Date() / 1000) - 1000
-        ensureAuthenticated({ provider })(req, res, nextSpy)
+      })
+
+      context('method is GET', function() {
+        beforeEach(function() {
+          req.method = 'GET'
+          ensureAuthenticated({ provider })(req, res, nextSpy)
+        })
+
+        it('should not call next', function() {
+          expect(nextSpy).not.to.be.called
+        })
+
+        it('should redirect to the authentication URL', function() {
+          expect(res.redirect).to.be.calledWith(`/connect/${provider}`)
+        })
+
+        it('should set the redirect URL in the session', function() {
+          expect(req.session.originalRequestUrl).to.equal('/test')
+        })
+      })
+
+      context('method is standard POST', function() {
+        beforeEach(function() {
+          req.method = 'POST'
+          req.body = { foo: 'bar' }
+          ensureAuthenticated({ provider })(req, res, nextSpy)
+        })
+
+        it('should set originalRequestBody property on session', function() {
+          expect(req.session.originalRequestBody).to.deep.equal({ foo: 'bar' })
+        })
+
+        it('should not call next', function() {
+          expect(nextSpy).not.to.be.called
+        })
+
+        it('should redirect to the authentication URL', function() {
+          expect(res.redirect).to.be.calledWith(`/connect/${provider}`)
+        })
+
+        it('should set the redirect URL in the session', function() {
+          expect(req.session.originalRequestUrl).to.equal('/test')
+        })
+      })
+
+      context('method is multipart upload', function() {
+        beforeEach(function() {
+          req.method = 'POST'
+          req.header = sinon.stub()
```

Comment:
You could define the stubs in the parent `beforeEach`. Then in this block where you just want to change what it returns you only have to do:

```javascript
...
req.header.returns('multipart/form-data; boundary=xxxx')
req.t.returns('multipartErrorString')
...
```

Reply:
OK